### PR TITLE
Use new go-state-types state accessors

### PR DESCRIPTION
**Pull request by arajasek**
This rips out the state types from specs-actors, but not any of the actors logic. The goal is for this to be used by Go-based implementations of the Filecoin protocol that need convenient accessors for the state types and method parameters of the Filecoin builtin actors. 

No part of this is intended to be used for message execution.

### CODE REVIEW

File: builtin/network.go
Comment (vyzo) [inline builtin/network.go:19]: how will this work for devnets etc?
Maybe we should have build tags?

File: builtin/sector.go
Comment (vyzo) [inline builtin/sector.go:15]: same problem exists here.

File: builtin/v8/manifest/manifest.go
Comment (vyzo) [inline builtin/v8/manifest/manifest.go:1]: Maybe this should be outside the v8 structure?
We dont anticipate it changing, amd it is kind of self versioned.

File: builtin/v8/market/market_balances.go
Comment (vyzo) [inline builtin/v8/market/market_balances.go:3]: delete?

File: builtin/v8/market/policy.go
Comment (vyzo) [inline builtin/v8/market/policy.go:10]: the usual problem.

**Review by vyzo — COMMENTED** [review]
Overall this looks good, and i am om board.

A couple of issues:
- the manifest type (and also, adt) should probably live outsied the v8 dir.
- bigger problem is the constants. Maybe we should make policy objects grouping them, similar to what we do in builtin-actors?

**Comment by anorth** [comment]
Here is the venue for design discussion: https://github.com/filecoin-project/lotus/issues/8607

**Review by anorth — COMMENTED** [review]
I think everything needs to go under the `v8` package, including `builtin`. Those are liable to change between versions too. Consumers like Lotus already had to deal with versioning of these.

File: builtin/network.go
Comment (anorth) [reply builtin/network.go:19]: Whatever was already working here should keep working. Let's not bring in new scope to this PR which should be essentially a mechanical move/rename.

File: builtin/v8/manifest/manifest.go
Comment (anorth) [inline builtin/v8/manifest/manifest.go:15]: This has no documentation so I have no idea what it's for (probably already an issue before you moved it).

File: builtin/v8/market/market_params.go
Comment (anorth) [inline builtin/v8/market/market_params.go:28]: It looks like you dropped the first line of this comment ("changes since version xxx")

File: builtin/v8/market/market_params.go
Comment (anorth) [inline builtin/v8/market/market_params.go:75]: I can see why you retained this, but I'm not sure it belongs here. This is really just actor logic that happened to be the state file for some reason. While maintaining two language versions of the state schemas is reasonable, maintaining duplicate application logic is not.

This was exported for use by nodes, but in in the future a better approach to things like this will be to call a readonly/view method on an actor locally in order to use its logic. The market actor may even become non-builtin in the future, other market actors certainly won't be.

I don't want to block on this, but this approach won't be sustainable.

This belongs in state, rather than params, I think.

File: builtin/v8/market/market_state.go
Comment (anorth) [inline builtin/v8/market/market_state.go:105]: You shouldn't need this exported, it's an internal detail

File: builtin/v8/market/policy.go
Comment (anorth) [reply builtin/v8/market/policy.go:10]: These aren't really state, though I see why you wanted to export them similarly. Again, long term, we probably want the actors to encapsulate their logic/policy better, and access it via view methods.

File: builtin/v8/miner/bitfield_queue.go
Comment (anorth) [inline builtin/v8/miner/bitfield_queue.go:32]: These are an interesting case. There is logic here tightly coupled to the actor implementations. You've taken all of the mutation methods as well as inspection. It would be really annoying for future Go actor development if this logic were separate from the main actor logic like this – so it's a good think we're not doing any!

One possible justification for including this here is for future migrations (but the long term plan for that should be that they're written in Rust).

For now, I think I would suggest excluding all the mutation methods, and keeping this repo as thin as possible. Is there a case I'm missing that motivates including them?

File: builtin/v8/miner/deadline_assignment.go
Comment (anorth) [inline builtin/v8/miner/deadline_assignment.go:14]: This is logic, not state, it doesn't belong in this repo.

File: builtin/v8/miner/deadline_state.go
Comment (anorth) [inline builtin/v8/miner/deadline_state.go:148]: I don't think including all the mutation methods in this repo is appropriate. It's a very large amount of essentially application logic. There's no way  we'll keep it in sync with the canonical actor code.

File: builtin/v8/miner/expiration_queue.go
Comment (anorth) [inline builtin/v8/miner/expiration_queue.go:47]: Application code that shoulnd't be here.

File: builtin/v8/miner/miner_state.go
Comment (anorth) [inline builtin/v8/miner/miner_state.go:313]: Lots of application logic in this file.

File: builtin/v8/miner/monies.go
Comment (anorth) [inline builtin/v8/miner/monies.go:13]: This is mostly logic, not state

**Review by anorth — COMMENTED** [review]
I'm down with this approach in general, but this PR is currently extracting way too much stuff. For the more complex actors, it's pulled out a huge amount of application logic that happens to be implemented on the state objects. This is actor code, mostly unnecessary for interpreting the state, and will be infeasible to keep in sync with the real version that counts, in builtin-actors. I didn't comment on all the occurrences of this, but there's a load in miner actor and others.


Since essentially everything in this repo represents a duplication of an idea expressed elsewhere, we should strive really hard to keep it as minimal as possible.

**Review by anorth — COMMENTED** [review]
Reward testdata files don't belong here. Nor util/math, util/smoothing

File: builtin/network.go
Comment (ZenGround0) [reply builtin/network.go:19]: Strong agree with @anorth

**Comment by arajasek** [comment]
From sync with @ZenGround0 : we should move the cbor generation into the root gen.go (and get it to generate cbor for all versions).

File: builtin/v8/market/market_params.go
Comment (arajasek) [reply builtin/v8/market/market_params.go:75]: Yeah, you're quite that this is going to be unsustainable / will eventually diverge from the real source of truth (builtin actors). We'll open an issue for it.

Agreed that it should be in state, not params, though. Making that change now.

File: builtin/v8/miner/miner_types.go
Comment (ZenGround0) [inline builtin/v8/miner/miner_types.go:247]: please move these guys out of v8

**Review by ZenGround0 — APPROVED** [review]
LGTM after moving proofs out of v8

**Comment by ZenGround0** [comment]
LGTM

**Review by anorth — COMMENTED** [review]
Definitely an improvement, thanks. 
- I think I understand why they're still there, but I'd love to see as much of the `policy.go` (and related `monies.go` etc) code removed if possible.
- Any chance of removing (part of) `reward_logic.go`?
- Ditto `util/smoothing/alpha_beta_filter.go`?

**Comment by arajasek** [comment]
@anorth Unfortunately, I think everything left behind is used by Lotus today:

- The remaining `policy` and `monies` is needed for making decisions about deal collateral, how much to send with Pre/Prove commit messages, etc.
- `reward_logic` (what's left) is currently needed to construct the reward state, which is needed in order to be able to start new networks.
- `alpha_beta_filter` is used when performing calculations about `monies`.

**Pull request merged**
